### ISSUE TITLE
Support configurable custom Identity Providers

### ISSUE THREAD

**Issue by stianst** [keycloak/keycloak#15344]
### Description

Identity Providers are not using `ConfiguredProvider`, but rather something bespoke, resulting in the ability to make configurable custom identity providers without extending the admin console.

The old admin console (somewhat of a hack) supported adding a custom template for configuring custom providers, but this is not supported in the new admin console, resulting in it not being possible to create custom identity providers and make them configurable through the new admin console.

We will have to refactor identity providers to be `ConfiguredProviders` and also update the new admin console to support configuring them like it does for other `ConfiguredProviders`.

This will have to be done prior to removing the old admin console, as otherwise we have lost functionality in the new admin console.

**Comment by stianst** [comment]
@ssilvert @mposolda @pedroigor FIY

**Comment by stianst** [comment]
Having community maintained extensions to support additional providers (AppleID, WebEx, etc.) that we are not able to support ourselves is very important.

**Comment by mposolda** [comment]
+1 that we need to support this.

Few additional thoughts:
- There is already supported to add new provider implementation though. At least when I run Keycloak with embedded undertow (KeycloakServer), I can see additional identity provider in the admin console, which is available just in the testsuite (TestKeycloakOidcIdentityProviderFactory) . However it seems to have only 4 hardcoded configuration fields available. So this should be improved.
 
- Also we can make sure that if IdentityProvider inherits from `SocialIdentityProvider`, it would be displayed in the `Social` section instead of `User-defined` section in the Identity providers list.

- There is some flexibility needed when rendering current pages for OpenID Connect and SAML identity providers. For example admin can check "Use discovery endpoint" checkbox and based on that, he can dynamically download configuration from OIDC and other similar things. This won't be easily possible if rendering is based just on the fields configured by `ConfiguredProvider` . So it looks that pages for OIDC and SAML providers will still need to be adjusted to NOT use the `ConfiguredProvider` mechanisms. But it looks that this is possible as same problem exists for user-storage provider and LDAP (LDAP page is also more flexible than other pages of regular user-provider, which are rendered dynamically).

**Comment by stianst** [comment]
> +1 that we need to support this.
> 
> Few additional thoughts:
> 
> * There is already supported to add new provider implementation though. At least when I run Keycloak with embedded undertow (KeycloakServer), I can see additional identity provider in the admin console, which is available just in the testsuite (TestKeycloakOidcIdentityProviderFactory) . However it seems to have only 4 hardcoded configuration fields available. So this should be improved.
> * Also we can make sure that if IdentityProvider inherits from `SocialIdentityProvider`, it would be displayed in the `Social` section instead of `User-defined` section in the Identity providers list.
> * There is some flexibility needed when rendering current pages for OpenID Connect and SAML identity providers. For example admin can check "Use discovery endpoint" checkbox and based on that, he can dynamically download configuration from OIDC and other similar things. This won't be easily possible if rendering is based just on the fields configured by `ConfiguredProvider` . So it looks that pages for OIDC and SAML providers will still need to be adjusted to NOT use the `ConfiguredProvider` mechanisms. But it looks that this is possible as same problem exists for user-storage provider and LDAP (LDAP page is also more flexible than other pages of regular user-provider, which are rendered dynamically).

We might need some form of "dynamic" actions or something to support those use-cases fully. We have a similar issue for custom user federation providers, where there's also some "hardcoded" actions for LDAP provider for example.

**Comment by ssilvert** [comment]
As a replacement for the old custom templates, there is the opportunity to reuse much of what we have done for User Profile.  Someone who creates a custom identity provider with custom attributes could also create the metadata we need to dynamically render the proper form.  The metadata could be created using JSON or an online configurator just like what we have in `Realm settings --> User Profile`.

The next step is to create a renderer that can render the form in React.  Erik is already working on the renderer for User Profile that will be used in both admin console and account console.  It wouldn't be much effort to reuse that for custom identity providers.

**Comment by stianst** [comment]
> As a replacement for the old custom templates, there is the opportunity to reuse much of what we have done for User Profile. Someone who creates a custom identity provider with custom attributes could also create the metadata we need to dynamically render the proper form. The metadata could be created using JSON or an online configurator just like what we have in `Realm settings --> User Profile`.
> 
> The next step is to create a renderer that can render the form in React. Erik is already working on the renderer for User Profile that will be used in both admin console and account console. It wouldn't be much effort to reuse that for custom identity providers.

The way we do custom config for authenticators and mappers (configured provider) is more suitable for identity providers than what's done for user profile (and client policies).

**Comment by nicolabeghin** [comment]
Following

**Comment by stianst** [comment]
@ammmze I'm afraid we had to remove the old admin console as we simply couldn't continue maintaining it, and there was an increase number of CVEs due to relying on EOL Angular 1.x. Since we have not really ever supported extending the admin console to add custom pages for custom identity providers, this was done in an unsupported way in the past, as such after reconsidering I don't consider it a breaking change. I know that doesn't help you all that much though.

It's unlikely that the Keycloak team will be able to work on this for Keycloak 22, and maybe not even for Keycloak 23 I'm afraid. A community contribution would be very welcome, as this is a reasonable big problem as a fair amount of people such as yourself are relying on custom identity providers.

**Comment by ammmze** [comment]
Thank you for that detail. I am not too surprised, I knew going in that really doing anything with custom identity providers is not officially supported (as the logs indicate when we implement the internal provider). So honestly I was surprised when it was mentioned to be a blocking change initially.

**Comment by mduchrow** [comment]
This is a severe problem. Two companies I am supporting are relying on custom identity providers. For them it is a blocker, they cannot upgrade to KC 21.
Is somebody already working on this? Can I help? What's the plan? Making identity providers `ConfiguredProvider` as Stain suggested?

**Comment by GVincenzo** [comment]
It's blocking for my company as well, and we've added custom pages in the new admin console for us to work with.
We had to manage new paths and static pages above the admin console application and build all the project

**Comment by antikalk** [comment]
@GVincenzo You had to fork the admin-ui right?

@mduchrow It would make sense to make it similar like for a custom `Authenticator`. These are provided via the `AuthenticatorFactory` which is a `ConfiguredProvider`. In case of the new UI custom Authenticators are already picked up by the Frontend and the UI rendered accordingly. So most of the implementation for custom identity providers could be oriented on that I think.

**Comment by GVincenzo** [comment]
> @GVincenzo You had to fork the admin-ui right?
> 
> @mduchrow It would make sense to make it similar like for a custom `Authenticator`. These are provided via the `AuthenticatorFactory` which is a `ConfiguredProvider`. In case of the new UI custom Authenticators are already picked up by the Frontend and the UI rendered accordingly. So most of the implementation for custom identity providers could be oriented on that I think.

yes I forked to an internal corporate repo and I actually have two remotes to periodically merge the official admin UI

**Comment by mduchrow** [comment]
@antikalk Yes, doing it like Authenticator is a good approach. 
However, there's much more to be considered. The `ConfiguredProvider` plugins are storing their data in `ComponentModel` (i.e. tables COMPONENT and COMPONENT_CONFIG) whereas `IdentityProvider` plugins are using `IdentityProviderModel` (i.e. tables IDENTITY_PROVIDER and IDENTITY_PROVIDER_CONFIG) with 10 existing subclasses.
So the real challenge is to migrate identity providers to `ComponentModel`. That includes moving all existing identity provider's config data in the database to COMPONENT and COMPONENT_CONFIG (using liquibase). Not only for custom identity providers, but also for all built-in identity providers.
And of course without _breaking changes_ in the code.

**Comment by jonkoops** [comment]
@stianst @ssilvert should we start prioritizing this now that the new UI is the default? Seems like plugin developers are experiencing issues with this (see https://github.com/keycloak/keycloak/discussions/19356).

**Comment by Romain7495** [comment]
Is it still planned for 22 milestone. Important, because some of our clients cannot upgrade to keycloak 21 without this feature

**Comment by ssilvert** [comment]
> Is it still planned for 22 milestone? Important, because some of our clients cannot upgrade to keycloak 21 without this feature

We have started working on a solution.  Not sure if it will make Keycloak 22 or not.

**Comment by danielFesenmeyer** [comment]
It would also be great if the compatible mappers could somehow "configured" for custom identity providers, see https://github.com/keycloak/keycloak/issues/21154

**Comment by jonkoops** [comment]
@danielFesenmeyer would #21097 not accomplish that? Feel free to provide input on the PR.

**Comment by danielFesenmeyer** [comment]
@jonkoops Looks like it doesn't. https://github.com/keycloak/keycloak/issues/21154 is a topic **not** directly related to the new Admin UI, and probably requires some discussion.

Just linked it here, because it's also related to the configuration of Identity Providers. It's probably better to handle that in a separate PR - if it's a common use case to re-use existing mappers for custom Identity Providers.

**Comment by ullgren** [comment]
@ssilvert  Seems the solution limits this to "Social" identity providers (registered through SocialIdentityProviderFactory) . However in the case of the Swedish BankId provider we do not want to inherit the social fields and limitations (such as only being able to create one per realm). Instead we would like to have the same option to create a configurable custom "User defined" identity provider (registered through the IdentityProviderFactory service).

Is this just an oversight in the fix of this issue or is this a completely new feature request ?

Also if the above question does not makes sense let me know and I can provide a simple example.

**Comment by danielFesenmeyer** [comment]
We are currently implementing an Azure AD IDP provider for Keycloak, because some Azure AD functionality (like reading transitive group membership) is not possible with the Keycloak builtin OIDC provider (OIDCIdentityProvider). We would like to have all the configuration possibilities for OIDCIdentityProvider, with some additions - and not the SocialIdentityProvider restrictions above.

Thus it would be important that additional properties could be configured via the UI, such as a filter for the groups to be loaded, or which AD field (id, displayName, ...) should be used as group identifier.

**Comment by edewit** [comment]
@danielFesenmeyer so why can't you use this solotion to add those fields?

**Comment by ullgren** [comment]
I will not answer for @danielFesenmeyer but in my case the problem is that the current  solution requires you to implement `SocialIdentityProvider` which will automatically add `Client id` and `Client secret` as mandatory configuration fields in the Admin UI.

It also blocks you from adding multiple Providers of the same type in the same realm, for SocialIdentityProvider you can only have one.

In my case, the Swedish BankId Provider, there is no Client id/secret to add since it is based on a completely different API not OAuth or anything like it.
So having these extra fields just makes the user experience weird. My use case also requires us to be able to have multiple provides of the same type in the same realm.
Granted I could work around that using multiple realms and federate them.

**Comment by edewit** [comment]
@ullgren right that would be needed, but wouldn't also mean that the interface needs to me moved higher up in the hierarchy

**Comment by ullgren** [comment]
@edewit You are probably correct. As I wrote I'm not a react/typescript/front end developer, I know just enough to make a mess. But if nobody beats me to it I will have a stab at it and see if I can get it to work.

The limitation of just being for social identity providers aside I really like the way this was implemented using the `ConfiguredProvider`. It very good devx to be able to add properties in java code using the builder pattern. :1st_place_medal: to all involved.

**Comment by evan10s** [comment]
Hi @edewit, I noticed that your change fixes non-social IdentityProviders not appearing on the Add Provider page but not on the details page for existing providers. I think this would be a similar change in`js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx` to what you implemented in #21320.

I'm happy to open a PR to fix, but I thought I'd let you know in case you're in a position to make the change faster than me.

**Comment by talbert83** [comment]
Could someone please point me to some documentation about how to add custom attributes to Identity Providers after all these changes? I don't think that going through multiple discussion threads and PRs is the right way to find out what one needs to do.

Our Keycloak extensions rely on custom IdP properties that we could previously view & edit in the Admin UI and we're trying to figure out what to change to be able to upgrade to Keycloak 23. Thanks!

**Comment by nicolabeghin** [comment]
> Could someone please point me to some documentation about how to add custom attributes to Identity Providers after all these changes? I don't think that going through multiple discussion threads and PRs is the right way to find out what one needs to do.

if useful, here's a WIP PR of mine where this is being done https://github.com/italia/spid-keycloak-provider/pull/59/commits/630137873f795f704b48b0cff652bac59015bde7 - please note: I still don't know how to show OOTB SAML/OIDC fields, open (but deserted :D) discussion in https://github.com/keycloak/keycloak/discussions/25862

**Comment by Jebarajjacob1** [comment]
Hi , Do you what authentication Nafath is using SAML or OAuth token ? If not what authentication they are using like cookies.

**Comment by theodol59** [comment]
Good afternoon,

Could somebody tell me how to reach custom config fields in a class, which extends AbstractIdentityProvider?

I have created custom IDP provider, which extends AbstractIdentityProvider and it uses some special keys.. I hardcode them in my code now, but is so ugly... I want these keys to be editable in UI.
I was able to extend UI form in "Provider details" -> "General Settings" form by overriding AbstractIdentityProviderFactory's getConfigProperties() method and fields show up in the form, but... I am not able to access them in a class, which extends AbstractIdentityProvider, constructor... I see only clientSecret and clientId values in "config" dictionary there... No other values :/

Can somebody shed some light? Any examples how to do what I need? Or is is still a bug, not implemented in Keycloak?..
Please..
